### PR TITLE
Pruning with groups 

### PR DESCRIPTION
**Pull request by mkaglins**
- Added grouping of nodes that should be pruned together
- All nodes in the group pruned and masked together (sum of nodes is used as group norm)
- All model analysis was moved to the initialization phase: grouping dependent node, note which nodes we cannot prune

@AlexKoff88 @vshampor @vanyalzr Please, take a look.

### CODE REVIEW

**Comment by AlexKoff88** [comment]
A general comment. After this PR is merged we should consider retraining/tuning all the existing pruned models that we expose as checkpoints right now.

File: nncf/pruning/model_analysis.py
Comment (AlexKoff88) [inline nncf/pruning/model_analysis.py:50]: Did't get why it is not in any cluster?

File: nncf/pruning/model_analysis.py
Comment (AlexKoff88) [inline nncf/pruning/model_analysis.py:97]: Need a description of what this function does. I guess you meant the word "unite" in the beginning. "merge" is better in this case.

File: nncf/pruning/model_analysis.py
Comment (AlexKoff88) [inline nncf/pruning/model_analysis.py:122]: `cluster_special_ops_in_model` -> `cluster_special_ops`

File: nncf/pruning/model_analysis.py
Comment (AlexKoff88) [inline nncf/pruning/model_analysis.py:161]: `Analyse` -> `Analyze`. Analyze for what? Can you elaborate a little more in the description?

File: nncf/pruning/model_analysis.py
Comment (AlexKoff88) [inline nncf/pruning/model_analysis.py:171]: Strange name of the method. A short description is required.

File: nncf/pruning/utils.py
Comment (AlexKoff88) [inline nncf/pruning/utils.py:215]: Strange name again. Please use something like "nonprunable", etc.

File: nncf/dynamic_graph/graph.py
Comment (vshampor) [reply nncf/dynamic_graph/graph.py:769]: Add a type hint for `type_list` as well

File: nncf/pruning/model_analysis.py
Comment (vshampor) [inline nncf/pruning/model_analysis.py:23]: Method type hints are missing throughout this file, please add these accordingly.

File: nncf/pruning/model_analysis.py
Comment (vshampor) [inline nncf/pruning/model_analysis.py:157]: Perhaps `unit_clusters_for_nodes` and `cluster_special_ops_in_model` could be made instance methods of a class like `PruningAnalyzer`?

File: tests/pruning/test_model_pruning_analysis.py
Comment (vshampor) [inline tests/pruning/test_model_pruning_analysis.py:20]: Please consider making your test case definitions more explicit by using a dedicated struct with descriptive fields - review `tests\test_input_management.py` for an example of such struct usage.

File: nncf/pruning/model_analysis.py
Comment (vshampor) [inline nncf/pruning/model_analysis.py:243]: I assume that the final groups to be pruned are derived as a result of the interaction between the `Analyzer`, `Clusterization`, `NodesCluster` entities. There is quite a lot of code and logic here. I suggest that you unit-test the interaction, so that future maintainers could have some kind of a reference towards how this algorithm should work, and to avoid the situations in the field where the exportable pruning rate is less than the in-Pytorch pruning rate.

**Review by vshampor — CHANGES_REQUESTED** [review]

**Comment by AlexKoff88** [comment]
@asenina, I remember you did something for testing of the previous implementation of pruning. Can you please take a look at this PR as well from the validation standpoint?

File: nncf/pruning/base_algo.py
Comment (mkaglins) [reply nncf/pruning/base_algo.py:85]: No, it means algorithm params like `prune_downsample_conv`.

File: nncf/pruning/model_analysis.py
Comment (mkaglins) [reply nncf/pruning/model_analysis.py:161]: Added description.

File: nncf/pruning/model_analysis.py
Comment (mkaglins) [reply nncf/pruning/model_analysis.py:157]: I don't think so. `ModelAnalyzer` has a specific purpose: model analysis in order to mark potentially prunable paths in the model. Whereas `unit_clusters_for_nodes` and `cluster_special_ops_in_model` works with grouping of special (usually Eltwise nodes).

File: nncf/pruning/base_algo.py
Comment (vanyalzr) [inline nncf/pruning/base_algo.py:204]: Did you remove a check for frozen layers via the `ignore_frozen_layers` flag?

File: nncf/pruning/model_analysis.py
Comment (vanyalzr) [inline nncf/pruning/model_analysis.py:94]: None is returned anyway

Comment (mkaglins) [reply nncf/pruning/model_analysis.py:104]: Yep, but PEP-8 require consistency in such cases:
> Either all return statements in a function should return an expression, or none of them should.

File: nncf/pruning/base_algo.py
Comment (mkaglins) [reply nncf/pruning/base_algo.py:204]: No, it was of the problem of merging with the last develop. Fixed it.

File: nncf/pruning/model_analysis.py
Comment (mkaglins) [reply nncf/pruning/model_analysis.py:270]: I added unit-tests for all of the structures and functions that analyze the model and interaction of them when it is applicable. Please, check them in the `test_model_pruning_analysis.py` file.

**Review by vshampor — APPROVED** [review]

**Comment by vshampor** [comment]
@mkaglins please resolve conflicts and then we'll merge this.

**Pull request merged**
